XenAdmin/Wizards/PatchingWizard/PatchingWizard_AutomatedUpdatesPage.cs

```diff
@@ -290,68 +276,125 @@ private void UpdateStatusTextBox()
         private void WorkerDoWork(object sender, DoWorkEventArgs doWorkEventArgs)
         {
             var bgw = sender as UpdateProgressBackgroundWorker;
+            PlanAction action = null;
 
-            foreach (PlanAction action in bgw.AllActions)
+            try
             {
-                try
+                //running actions (non-delayed)
+                foreach (var a in bgw.PlanActions)
                 {
+                    action = a;
+
                     if (bgw.CancellationPending)
                     {
                         doWorkEventArgs.Cancel = true;
                         return;
                     }
 
-                    InitializePlanAction(bgw, action);
+                    RunPlanAction(bgw, action);
+                }
+
+                // running delayed actions, but skipping the ones that should be skipped
+                // iterating through hosts
+                foreach (var kvp in bgw.DelayedActionsByHost)
+                {
+                    var h = kvp.Key;
+                    var actions = kvp.Value;
+
+                    // iterating through delayed actions for this host
+                    foreach (var a in actions)
+                    {
+                        if (bgw.CancellationPending)
+                        {
+                            doWorkEventArgs.Cancel = true;
+                            return;
+                        }
 
-                    bgw.ReportProgress(0, action);
-                    action.Run();
+                        action = a;
 
-                    Thread.Sleep(1000);
+                        // always run the IAvoidRestartHostsAware ones (these are assumed to be preceding the rest, be careful if you touch that!)

```

Comment:
Instead of relying on the actions to be in a specific order in the list, we could iterate through the list twice:
- on first iteration, run only the IAvoidRestartHostsAware actions
- on second iteration run the other actions, if needed